### PR TITLE
Pairwise covariance

### PR DESCRIPTION
**Pull request by cristineguadelupe**

### CODE REVIEW

File: lib/explorer/data_frame.ex [hunk 1]
```diff
@@ -5659,6 +5659,7 @@ defmodule Explorer.DataFrame do
 
   @doc """
   Calculates the pairwise Pearson's correlation of numeric columns.
+  The returned dataframe is the correlation matrix.
```
Comment (philss) [inline lib/explorer/data_frame.ex:5662]:
```suggestion

  The returned dataframe is the correlation matrix.
```

File: lib/explorer/data_frame.ex [hunk 2]
```diff
@@ -5692,22 +5693,48 @@ defmodule Explorer.DataFrame do
   @spec correlation(df :: DataFrame.t(), opts :: Keyword.t()) :: df :: DataFrame.t()
   def correlation(df, opts \\ []) do
     opts = Keyword.validate!(opts, column_name: "names", columns: names(df), ddof: 1)
+    out_df = pairwised_df(df, opts)
+    Shared.apply_impl(df, :correlation, [out_df, opts[:ddof]])
+  end
 
-    column_name = to_column_name(opts[:column_name])
+  @doc """
+  Calculates the pairwise covariance of numeric columns.
```
Comment (philss) [inline lib/explorer/data_frame.ex:5701]:
```suggestion
  Calculates the pairwise covariance of numeric columns.

```

File: lib/explorer/data_frame.ex [hunk 3]
```diff
@@ -5692,22 +5693,48 @@ defmodule Explorer.DataFrame do
...
 
-  defp numeric_column?(df, name) do
-    Series.dtype(df[name]) in [:integer | Explorer.Shared.float_types()]
+  The resultant columns are always `{:f, 64}`.
+
+  ## Options
+
+  * `:columns` - the selection of columns to calculate. Defaults to all numeric columns.
+  * `:column_name` - the name of the column with column names. Defaults to "names".
+  * `:ddof` - the 'delta degrees of freedom' - the divisor used in the correlation
```
Comment (philss) [inline lib/explorer/data_frame.ex:5718]:
```suggestion
  * `:ddof` - the 'delta degrees of freedom' - the divisor used in the covariance
```

File: lib/explorer/polars_backend/data_frame.ex [hunk 4]
```diff
@@ -847,4 +837,25 @@ defmodule Explorer.PolarsBackend.DataFrame do
   def inspect(df, opts) do
     Explorer.Backend.DataFrame.inspect(df, "Polars", n_rows(df), opts)
   end
+
+  # helpers
+
+  defp pairwised(df, out_df, ddof, operation) do
+    [column_name | cols] = out_df.names
+
+    correlations =
```
Comment (philss) [inline lib/explorer/polars_backend/data_frame.ex:846]:
```suggestion
    pairwised_results =
```

File: lib/explorer/polars_backend/data_frame.ex [hunk 5]
```diff
@@ -847,4 +837,25 @@ defmodule Explorer.PolarsBackend.DataFrame do
...
+          |> Enum.map(fn right -> apply(PolarsSeries, operation, [df[left], df[right], ddof]) end)
+          |> Shared.from_list({:f, 64})
+          |> Shared.create_series()
+
+        {left, corr_series}
+      end)
+
+    names_series = cols |> Shared.from_list(:string) |> Shared.create_series()
+
+    from_series([{column_name, names_series} | correlations])
```
Comment (philss) [inline lib/explorer/polars_backend/data_frame.ex:859]:
```suggestion
    from_series([{column_name, names_series} | pairwised_results])
```

**Review by philss — COMMENTED** [review]
Just some minor points.

**Review by philss — APPROVED** [review]
:shipit:

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**